"javascript:" urls: execute in correct global scope

archive64 opened the pull request:
<!-- Please describe your changes on the following line: -->

#### Summary

This pull request makes `javascript:` urls execute in the correct global scope.

#### Example

```html
<script> var x = 4; </script>

<!-- this branch: logs "4" -->
<!-- master: undefined variable error -->
<a href="javascript:console.log(x)">link</a>
```

#### Questions

I'm new to servo and rust, so I'm unsure about these changes. In particular:
  * What's the appropriate place to evaluate the js?
    * I moved it to `handle_navigate`, but I'm not sure if this will catch all occurrences of `javascript:` urls. I also don't know if this will execute in the correct thread and the correct window.
  * What should I do with the result of the js evaluation?
    * I just ignored it. The previous behavior displayed it as the content of a new page load.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #15147, #16718

<!-- Either: -->
- [x] There are tests for these changes

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/17083)
<!-- Reviewable:end -->

archive64 committed:
"javascript:" urls: add web-platform-test

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @mbrubeck (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @fitzgen: components/script/script_thread.rs
 * @kichjang: components/script/script_thread.rs

jdm commented:
To answer your question about the result of execution, consider steps 9-11 of https://html.spec.whatwg.org/multipage/browsers.html#javascript-protocol. That's why the old code was treating it as the page content.

danielj41 commented:
Thank you!

I think I'll need to understand the spec more thoroughly before I can figure out the correct solution. I'll take another look at this when I get a chance.

danielj41 commented:
I'm having trouble getting this to work.

#### Summary

From the spec, handling `javascript` urls happens in the [navigate](https://html.spec.whatwg.org/multipage/browsers.html#navigate) algorithm, which I think corresponds to `script_thread.rs`'s `handle_navigate` function. That's where I moved the js evaluation code to in this pull request. (Though, it looks like I'm also supposed to queue a task from the "DOM manipulation task source," which I haven't done).

#### Problem

I don't know how to implement step 12.12: "Run process a navigate response [with the result of the js evaluation]." I can't figure out a good way to skip ahead to [process a navigate response](https://html.spec.whatwg.org/multipage/browsers.html#process-a-navigate-response), which I think corresponds to `script_thread.rs`'s `load`.

#### Attempted solution 1

The most similar example I could find was `start_page_load_about_blank`, which "simulates" a response. However, copying this code requires an `InProgressLoad` object, which I don't think is available when `handle_navigate` runs. Since I can't add that object to `self.incomplete_loads`, the `assert!` in `handle_page_headers_available` fails.

#### Attempted solution 2

I also tried running the js evaluation code in `pre_page_load`. At this point, the `InProgressLoad` object is available, but I wasn't able to access the `Window` associated with `load_data.creator_pipeline_id`. I think I need this window to run the js in the correct global. When I tried `find_window()`, it returned `None`.

I'd greatly appreciate some insight from someone more familiar with Servo. Thank you for your time! :)

jdm commented:
Sorry, this has been on my list of PRs that I want to give useful feedback on. I'll prioritize doing so if you're still interested in working on it!

bors-servo commented:
:umbrella: The latest upstream changes (presumably #17822) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
Sorry for the extreme delay here! I kept putting off responding because I thought it would take a lot of time to wrap my head around the various complexities here, but your last comment did a really good job of laying out the issues you're facing.

I agree that neither of your attempted solutions quite fit the existing model. Here is what I recommend instead:
* continue performing the JS evaluation in ScriptThread::handle_navigation (for the time being that's fine)
* store the resulting body in a new field in the LoadData structure
* in ScriptThread::handle_new_layout, obtain the body from the LoadData embedded in the NewLayoutInfo struct
* modify ScriptThread::start_page_load_about_blank to support passing in the desired response body from the caller
* add an extra check for javascript urls to ScriptThread::handle_new_layout that passes the body to the previously modified method

Does that make sense?

danielj41 commented:
Thank you for taking the time to look through it!

Yes, that makes sense to me. I'll work on this a bit more when I get a chance.

---

I do have one follow-up question about non-string results. This is what the spec says:

* **navigate 12.11:** Process result: If `Type(result)` is not `String`, then set `response` to a response whose status is 204.
* **process a navigate response 2:** If response's status is 204 or 205, then abort these steps.

However, I didn't see any code that aborts loading a page on a 204 status (but I may have overlooked it). Which implementation do you think would make more sense:

* In `ScriptThread::handle_navigate`, abort early if the result is not a string.
  * This seems simplest, but doesn't quite follow the spec.
* In `ScriptThread::handle_navigate`, pass a 204 status through with the `LoadData`, then abort at a later function.
  * I'm not sure where to do this before the next page starts loading.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #17425) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
I would recommend passing the 204 status through (we can have an enum that either represents a response body or a 204 response), and then ScriptThread::start_page_load_about_blank (or an equivalent method) can adjust the Metadata value if the status code needs to change.

archive64 committed:
"javascript:" urls: run in correct global

Make some changes to javascript scheme url evaluation to bring it
closer to
https://html.spec.whatwg.org/multipage/browsing-the-web.html#navigating-across-documents

- Evaluate the js before the page load so that it happens in the
  correct `window` global.
- If the result is not a string, the response should be a 204.

This required saving some data in load_data, since it's not
possible to modify the response at the point where we're evaluating
the js.

jdm commented on tests/wpt/web-platform-tests/url/a-element-href-javascript.html:
```diff
@@ -0,0 +1,28 @@
+
+document.querySelector("#javascript-link").click();
+
+step_timeout(function() {
```
Let's do this instead:
```html
<script>
function changeStatus() {
  t.done();
}

var t = async_test(function(t) {
  document.querySelector("#javascript-link").click();
}, "javascript: scheme urls should be executed in current global scope");
</script>
```

danielj41 replied:
Fixed in https://github.com/servo/servo/pull/17083/commits/fc23cb1a6306e830e622b93e6eccc13cf66bd0a4

jdm commented on components/script/script_thread.rs:
```diff
@@ -2327,16 +2307,64 @@ impl ScriptThread {
             None => {
+                let is_javascript = load_data.url.scheme() == "javascript";
+                if is_javascript {
+                    use url::percent_encoding::percent_decode;
```
Move this to with the other use statements at the top of the file.

danielj41 replied:
Fixed in https://github.com/servo/servo/pull/17083/commits/5d28dd64d929ce471a7889a4bdddf8d1acb61184

jdm commented on components/script/script_thread.rs:
```diff
@@ -2327,16 +2307,64 @@ impl ScriptThread {
                     iframe.navigate_or_reload_child_browsing_context(Some(load_data), NavigationType::Regular, replace);
                 }
+
+                // TODO: Test javascript: urls in iframes
```
This looks more like a "note to self"; want to file an issue for this and remove the comment instead?

danielj41 replied:
I was planning on looking into it (and [this `TODO`]( https://github.com/servo/servo/pull/17083/commits/fa3e9ab2444daab2c67b7e25fd6a5f88264b4ddb#diff-1879ac6bd5d567e2aa43529d33474677R1685)) a bit more in this pull request. I'll open new issues if they aren't easy to fix!

danielj41 replied:
I added support for `<iframe src="javascript:'something'"></iframe>` in https://github.com/servo/servo/pull/17083/commits/6ae6031468662161280a4353da9c0c23368cd83d. This was necessary to prevent a regression from `master`.

I fixed the other `TODO` in https://github.com/servo/servo/pull/17083/commits/ff786a050a9b8c8f526130e4bca9b2208f7998c4.

jdm commented on components/script/script_thread.rs:
```diff
@@ -2327,16 +2307,64 @@ impl ScriptThread {
+                }
+
+                if is_javascript {
+                    load_data.url = ServoUrl::parse("about:blank").unwrap();
```
This block can be merged with the previous one.

danielj41 replied:
I tried to make this look cleaner in https://github.com/servo/servo/pull/17083/commits/5d28dd64d929ce471a7889a4bdddf8d1acb61184. I couldn't merge it with the previous block there was an immutable reference to `url` in that block.

jdm reviewed (changes requested):
This looks really great! Please be sure to check that the new test passes in Firefox too; if the test file resides in tests/wpt/web-platform-tests/html/something you can visit `http://localhost:8000/html/browsers/navigating-across-documents/html/something` in another browser while the test window is still open in Servo.

archive64 committed:
"javascript:" urls: clean up after aborting a page load

Problem:
  After aborting on a 204 or 205 status code, you could no longer
  follow links on the page.

Cause:
  `constellation.rs` ignores new LoadUrl requests since the aborted
  one is in its `pending_changes` list.

Solution:
  Send a message to constellation that lets it clean up its
  `pending_changes` list.

archive64 committed:
"javascript:" urls: move test to correct location

- move test to correct directory
- update it to be more concise

I also confirmed that this test passes in Firefox.

archive64 committed:
"javascript:" urls: clean up js evaluation code

- move it to its own function
- move the `url = "about:blank" code into the same block
- move the `use` statement to the top of the file

danielj41 commented on components/script/script_thread.rs:
```diff
@@ -2309,69 +2310,70 @@ impl ScriptThread {
         }
     }
 
+    fn eval_js_url(&self, pipeline_id: PipelineId, load_data: &mut LoadData) {
```
I'm noticing that this is failing:

```
tests/wpt/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/javascript-url-return-value-handling.html
```

which tries:

```html
<iframe src="javascript:'1'"></iframe>
```

I think I need to call this `eval_js_url` from somewhere in `htmliframeelement.rs`. I'll look into this more soon!

archive64 committed:
"javascript:" urls: evaluate in iframe src attribute

- generalize the eval_js_url function so it can be called from
  multiple places.
- call it in htmliframeelement.
- if the js eval results in a non-string result, then it won't
  navigate to a new page, so don't block on the new page loading.

danielj41 commented on tests/wpt/web-platform-tests/url/a-element-href-javascript.html:
```diff
@@ -0,0 +1,28 @@
+<!doctype html>
```
Thanks! Fixed in https://github.com/servo/servo/pull/17083/commits/fc23cb1a6306e830e622b93e6eccc13cf66bd0a4

jdm commented:
Is this ready for re-review?

danielj41 commented:
Yes, thank you!

jdm commented on components/script/script_thread.rs:
```diff
@@ -2309,69 +2310,70 @@ impl ScriptThread {
+                    Some(JsEvalResult::NoContent)
+                };
+            }
+        };
```
I don't think this surrounding {} is necessary any more.

danielj41 replied:
Removed the block in https://github.com/servo/servo/pull/17083/commits/709cd3a0b636f601c0a4ca606c241500eca13789?w=1, does that look ok?

jdm reviewed (changes requested):
Only small comments left! This looks like a nice improvement; thanks for working on it!

archive64 committed:
"javascript:" urls: remove unnecessary block

Remove this block and unindent the code one level.

Doing this required cloning `load_data.url` so that we could
later mutate `load_data.url` in the same block.

danielj41 commented on components/script/script_thread.rs:
```diff
@@ -1676,19 +1676,19 @@ impl ScriptThread {
                     _ => ()
                 };
 
+                let load = self.incomplete_loads.borrow_mut().remove(idx);
```
If I removed the `load` from `incomplete_loads` here, it leaves the pipeline in an invalid state:
 * the load is "complete"
 * but, the pipeline does not have a `Document`

Concretely, this is a problem when `handle_exit_pipeline_msg` gets called. It gets called as a result of sending `ScriptMsg::AbortLoadUrl` here. It has a [block of code](https://github.com/danielj41/servo/blob/709cd3a0b636f601c0a4ca606c241500eca13789/components/script/script_thread.rs#L1767) that looks like this pseudocode:

```
if (pipeline in incomplete_loads) {
    // do cleanup
} else if (pipeline has document) {
    // do cleanup
} else {
    warn! (pipeline doesn't exist)
}
```

jdm replied:
Oh, the fact that it gets cleaned up in handle_exit_pipeline_msg makes me feel better, in that case.

danielj41 commented:
Thank you for the review! This is ready to be reviewed again.

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 709cd3a has been approved by `jdm`

<!-- @bors-servo r=jdm 709cd3a0b636f601c0a4ca606c241500eca13789 -->

bors-servo commented:
:hourglass: Testing commit 709cd3a0b636f601c0a4ca606c241500eca13789 with merge b8bec1200e49efaa6fe031b4efdaf4f72ce27553...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5966)

jdm commented:
```
  ▶ OK [expected TIMEOUT] /XMLHttpRequest/open-url-javascript-window-2.htm

  ▶ Unexpected subtest result in /XMLHttpRequest/open-url-javascript-window-2.htm:
  │ FAIL [expected TIMEOUT] XMLHttpRequest: open() - resolving URLs (javascript: <iframe>; 2)
  └   → The string did not match the expected pattern.

  /XMLHttpRequest/open-url-multi-window-3.htm
  ▶ OK [expected TIMEOUT] /XMLHttpRequest/open-url-javascript-window.htm

  ▶ Unexpected subtest result in /XMLHttpRequest/open-url-javascript-window.htm:
  └ PASS [expected TIMEOUT] XMLHttpRequest: open() - resolving URLs (javascript: <iframe>; 1)

  ▶ Unexpected subtest result in /dom/nodes/Document-contentType/contentType/contenttype_javascripturi.html:
  └ PASS [expected FAIL] Javascript URI document.contentType === 'text/html'

  ▶ Unexpected subtest result in /html/browsers/browsing-the-web/navigating-across-documents/013.html:
  └ PASS [expected FAIL] Link with onclick navigation to javascript url with delayed document.write and href navigation 

  ▶ Unexpected subtest result in /html/browsers/browsing-the-web/navigating-across-documents/014.html:
  └ PASS [expected FAIL]  Link with javascript onclick form submission script order 

  /fetch/http-cache/status.html
  ▶ Unexpected subtest result in /html/browsers/browsing-the-web/navigating-across-documents/015.html:
  └ PASS [expected FAIL]  Link with javascript onclick and href script order 

  /html/browsers/browsing-the-web/navigating-across-documents/javascript-url-global-scope.html
  ▶ Unexpected subtest result in /html/browsers/browsing-the-web/navigating-across-documents/javascript-url-query-fragment-components.html:
  └ PASS [expected FAIL] iframes with javascript src

  ▶ Unexpected subtest result in /html/dom/documents/dom-tree-accessors/Document.currentScript.html:
  └ PASS [expected NOTRUN] Script iframe-src

  ▶ Unexpected subtest result in /old-tests/submission/Opera/script_scheduling/028.html:
  └ PASS [expected FAIL]  scheduler: javascript: URL

  ▶ Unexpected subtest result in /old-tests/submission/Opera/script_scheduling/029.html:
  └ PASS [expected FAIL]  scheduler: javascript: URL in HREF
```
Looks like a bunch of test expectations can be updated to reflect the fact that they now pass (or at least don't timeout any longer).

archive64 committed:
"javascript:" urls: update test expectations

KiChjang commented:
By default, WPTs are expected to PASS or OK, so if there is an expectation INI file that expects only one test to FAIL before but is now passing, you may simply remove it.

danielj41 commented:
@KiChjang That makes more sense, thanks! I amended the commit.

KiChjang commented:
@bors-servo r=jdm

bors-servo commented:
:pushpin: Commit 452db05 has been approved by `jdm`

<!-- @bors-servo r=jdm 452db052ff385e7abd112dec3b8147199b5d5b5f -->

bors-servo commented:
:hourglass: Testing commit 452db052ff385e7abd112dec3b8147199b5d5b5f with merge 40c8a6389afdad8fdffdf4a5616627f9a51b1822...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/8799), [arm32](http://build.servo.org/builders/arm32/builds/9032), [arm64](http://build.servo.org/builders/arm64/builds/9063), [linux-dev](http://build.servo.org/builders/linux-dev/builds/9123), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5916), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5968), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/9164), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/1622), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/1629), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5859), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5916), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/1646), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/1641), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/4418)
Approved by: jdm
Pushing 40c8a6389afdad8fdffdf4a5616627f9a51b1822 to master...
